File: _old/internal/curlfile/curl.go
// Package curlfile imports/exports HTTP requests as curl one-liners.
//
// Import handles the common subset of curl flags people actually paste:
// -X/--request, -H/--header, -d/--data/--data-raw/--data-binary,
// --data-urlencode, -u/--user, --url, -F/--form (treated as multipart).
//
// As curl itself does, a request built from -d/--data* fields defaults to
// Content-Type: application/x-www-form-urlencoded unless one is set explicitly.
//
// Anything more exotic (--cert, --cacert, --resolve, --proxy, etc.) is silently
// dropped — the goal is best-effort, not full curl emulation.
package curlfile

import (
	"encoding/base64"
	"fmt"
	"net/url"
	"sort"
	"strings"

	"github.com/bgunnarsson/binman/internal/httpfile"
)

// IsCurl reports whether s looks like a curl invocation.
func IsCurl(s string) bool {
	t := strings.TrimSpace(s)
	return strings.HasPrefix(t, "curl ") || strings.HasPrefix(t, "curl\t") ||
		strings.HasPrefix(t, "curl\n") || t == "curl" || strings.HasPrefix(t, "$ curl ")
}

// Parse converts a curl command-line string into a Request.
func Parse(s string) (*httpfile.Request, error) {
	tokens, err := tokenize(s)
	if err != nil {
		return nil, err
	}
	if len(tokens) == 0 || (tokens[0] != "curl" && tokens[0] != "$") {
		return nil, fmt.Errorf("not a curl command")
	}
	// strip a leading "$ curl"
	if tokens[0] == "$" && len(tokens) > 1 && tokens[1] == "curl" {
		tokens = tokens[2:]
	} else {
		tokens = tokens[1:]
	}

	req := &httpfile.Request{Headers: map[string]string{}}
	method := ""
	var positional []string
	// Track whether any -d/--data/--data-urlencode field was seen so we can
	// mirror curl's implicit application/x-www-form-urlencoded content type.
	sawFormData := false

	for i := 0; i < len(tokens); i++ {
		tok := tokens[i]
		next := func() (string, bool) {
			if i+1 >= len(tokens) {
				return "", false
			}
			i++
			return tokens[i], true
		}
		switch {
		case tok == "-X" || tok == "--request":
			if v, ok := next(); ok {
				method = strings.ToUpper(v)
			}
		case tok == "-H" || tok == "--header":
			if v, ok := next(); ok {
				if idx := strings.Index(v, ":"); idx > 0 {
					req.Headers[strings.TrimSpace(v[:idx])] = strings.TrimSpace(v[idx+1:])
				}
			}
		case tok == "-d" || tok == "--data" || tok == "--data-raw" || tok == "--data-binary":
			if v, ok := next(); ok {
				if req.Body != "" {
					req.Body += "&"
				}
				req.Body += v
				sawFormData = true
				if method == "" {
					method = "POST"
				}
			}
		case tok == "--data-urlencode":
			if v, ok := next(); ok {
				if req.Body != "" {
					req.Body += "&"
				}
				if eq := strings.Index(v, "="); eq >= 0 {
					req.Body += v[:eq+1] + url.QueryEscape(v[eq+1:])
				} else {
					req.Body += url.QueryEscape(v)
				}
				sawFormData = true
				if method == "" {
					method = "POST"
				}
			}
		case tok == "-u" || tok == "--user":
			if v, ok := next(); ok {
				req.Headers["Authorization"] = "Basic " + base64.StdEncoding.EncodeToString([]byte(v))
			}
		case tok == "--url":
			if v, ok := next(); ok {
				req.URL = v
			}
		case tok == "-F" || tok == "--form":
			// Note as a header so the user knows; full multipart parse is out of scope.
			if v, ok := next(); ok {
				if req.Body != "" {
					req.Body += "&"
				}
				req.Body += v
				if method == "" {
					method = "POST"
				}
				if _, exists := req.Headers["Content-Type"]; !exists {
					req.Headers["Content-Type"] = "multipart/form-data"
				}
			}
		case strings.HasPrefix(tok, "-"):
			// Unknown flag with arg — try to swallow the next token if it doesn't look like a flag.
			if i+1 < len(tokens) && !strings.HasPrefix(tokens[i+1], "-") && !looksLikeURL(tokens[i+1]) {
				i++
			}
		default:
			positional = append(positional, tok)
		}
	}

	for _, p := range positional {
		if req.URL == "" {
			req.URL = p
		}
	}
	if method == "" {
		method = "GET"
	}
	req.Method = method
	// curl implicitly sends application/x-www-form-urlencoded for -d/--data
	// fields. Preserve that so form bodies (e.g. OAuth2 client_credentials token
	// requests) are recognized as forms and sent with the right content type.
	if sawFormData {
		if _, exists := req.Headers["Content-Type"]; !exists {
			req.Headers["Content-Type"] = "application/x-www-form-urlencoded"
		}
	}
	return req, nil
}

func looksLikeURL(s string) bool {
	return strings.HasPrefix(s, "http://") || strings.HasPrefix(s, "https://") ||
		strings.HasPrefix(s, "{{")
}

// Format renders req as a single-line curl command suitable for clipboard paste.
func Format(req *httpfile.Request) string {
	var b strings.Builder
	b.WriteString("curl")
	method := req.Method
	if method == "" {
		method = "GET"
	}
	if method != "GET" {
		b.WriteString(" -X ")
		b.WriteString(method)
	}
	keys := make([]string, 0, len(req.Headers))
	for k := range req.Headers {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		fmt.Fprintf(&b, " -H %s", shellQuote(k+": "+req.Headers[k]))
	}
	if req.Body != "" {
		fmt.Fprintf(&b, " --data-raw %s", shellQuote(req.Body))
	}
	if req.URL != "" {
		b.WriteByte(' ')
		b.WriteString(shellQuote(req.URL))
	}
	return b.String()
}

func shellQuote(s string) string {
	if !strings.ContainsAny(s, " \t\n'\"\\$`!*?[]()") {
		return s
	}
	return "'" + strings.ReplaceAll(s, "'", `'\''`) + "'"
}

// tokenize is a minimal POSIX-ish shell tokenizer that handles single quotes,
// double quotes, backslash-escapes, and line-continuation (`\` + newline).
// It is intentionally not complete — it covers what curl pastes look like.
func tokenize(s string) ([]string, error) {
	var (
		out      []string
		cur      strings.Builder
		inSingle bool
		inDouble bool
	)
	flush := func() {
		if cur.Len() > 0 {
			out = append(out, cur.String())
			cur.Reset()
		}
	}
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case inSingle:
			if c == '\'' {
				inSingle = false
				continue
			}
			cur.WriteByte(c)
		case inDouble:
			if c == '"' {
				inDouble = false
				continue
			}
			if c == '\\' && i+1 < len(s) {
				next := s[i+1]
				if next == '"' || next == '\\' || next == '$' || next == '`' || next == '\n' {
					cur.WriteByte(next)
					i++
					continue
				}
			}
			cur.WriteByte(c)
		default:
			switch c {
			case '\'':
				inSingle = true
			case '"':
				inDouble = true
			case '\\':
				if i+1 < len(s) {
					next := s[i+1]
					if next == '\n' {
						i++
						continue
					}
					cur.WriteByte(next)
					i++
				}
			case ' ', '\t', '\n', '\r':
				flush()
			default:
				cur.WriteByte(c)
			}
		}
	}
	if inSingle || inDouble {
		return nil, fmt.Errorf("unterminated quote")
	}
	flush()
	return out, nil
}

File: crates/binman-core/src/formats/curl.rs
//! curl command lines, in and out.
//!
//! Import covers what people actually paste, including what a browser's "Copy
//! as cURL" writes: `-X`, `-H`, the `-d`/`--data*` family, `--json`, `-u`,
//! `-F`, `-A`, `-e`, `-b`, `--url`, and `$'…'` quoting. Everything else curl
//! understands — `--cert`, `--resolve`, `--proxy` — is passed over: the aim is
//! the request, not curl's whole behaviour.
//!
//! As curl does, a body built from `-d` fields is sent as
//! `application/x-www-form-urlencoded` unless a Content-Type says otherwise.

use base64::Engine;

use crate::error::{Error, Result};
use crate::request::{Request, header, set_header};

/// Whether `text` is a curl invocation rather than a URL.
pub fn is_curl(text: &str) -> bool {
    let text = text.trim();
    text == "curl"
        || text.starts_with("curl ")
        || text.starts_with("curl\t")
        || text.starts_with("curl\n")
        || text.starts_with("$ curl ")
}

/// Flags that take a value.
const VALUED: &[&str] = &[
    "-X",
    "--request",
    "-H",
    "--header",
    "-d",
    "--data",
    "--data-raw",
    "--data-binary",
    "--data-ascii",
    "--data-urlencode",
    "--json",
    "-u",
    "--user",
    "--url",
    "-F",
    "--form",
    "--form-string",
    "-A",
    "--user-agent",
    "-e",
    "--referer",
    "-b",
    "--cookie",
];

/// Flags that take none, so the token after one is never mistaken for its
/// value.
const SWITCHES: &[&str] = &[
    "-s",
    "--silent",
    "-S",
    "--show-error",
    "-k",
    "--insecure",
    "-L",
    "--location",
    "-i",
    "--include",
    "-v",
    "--verbose",
    "-f",
    "--fail",
    "-N",
    "--no-buffer",
    "--compressed",
    "-#",
    "--progress-bar",
    "--http1.1",
    "--http2",
    "-g",
    "--globoff",
];

/// Short flags whose value may be written straight after them: `-XPOST`.
const ATTACHABLE: [&str; 5] = ["-X", "-H", "-d", "-u", "-F"];

pub fn parse(text: &str) -> Result<Request> {
    let tokens = tokenize(text)?;
    let mut rest = tokens.as_slice();
    if rest.first().is_some_and(|token| token == "$") {
        rest = &rest[1..];
    }
    match rest.first() {
        Some(first) if first == "curl" => rest = &rest[1..],
        _ => return Err(Error::Invalid("not a curl command".into())),
    }

    let mut request = Request::default();
    let mut method = None;
    let mut data: Vec<String> = Vec::new();
    let mut form_encoded = false;
    let mut multipart = false;

    let mut index = 0;
    while index < rest.len() {
        let token = rest[index].as_str();
        index += 1;

        let (flag, attached) = match ATTACHABLE
            .iter()
            .find(|flag| token.len() > flag.len() && token.starts_with(**flag))
        {
            Some(flag) => (*flag, Some(token[flag.len()..].to_string())),
            None => (token, None),
        };

        if VALUED.contains(&flag) {
            let value = match attached {
                Some(value) => value,
                None => match rest.get(index) {
                    Some(value) => {
                        index += 1;
                        value.clone()
                    }
                    None => break,
                },
            };
            match flag {
                "-X" | "--request" => method = Some(value.to_ascii_uppercase()),
                "-H" | "--header" => {
                    if let Some((name, value)) = value.split_once(':')
                        && !name.trim().is_empty()
                    {
                        request
                            .headers
                            .push((name.trim().to_string(), value.trim().to_string()));
                    }
                }
                "-d" | "--data" | "--data-raw" | "--data-binary" | "--data-ascii" => {
                    data.push(value);
                    form_encoded = true;
                }
                "--data-urlencode" => {
                    data.push(match value.split_once('=') {
                        Some((name, content)) => format!("{name}={}", query_escape(content)),
                        None => query_escape(&value),
                    });
                    form_encoded = true;
                }
                "--json" => {
                    data.push(value);
                    if header(&request.headers, "Content-Type").is_none() {
                        request
                            .headers
                            .push(("Content-Type".into(), "application/json".into()));
                    }
                    if header(&request.headers, "Accept").is_none() {
                        request
                            .headers
                            .push(("Accept".into(), "application/json".into()));
                    }
                }
                "-u" | "--user" => {
                    let encoded = base64::engine::general_purpose::STANDARD.encode(value);
                    set_header(
                        &mut request.headers,
                        "Authorization",
                        format!("Basic {encoded}"),
                    );
                }
                "--url" => request.url = value,
                "-F" | "--form" | "--form-string" => {
                    data.push(value);
                    multipart = true;
                }
                "-A" | "--user-agent" => set_header(&mut request.headers, "User-Agent", value),
                "-e" | "--referer" => set_header(&mut request.headers, "Referer", value),
                // Without an `=` it names a cookie file, which is not a request.
                "-b" | "--cookie" if value.contains('=') => {
                    set_header(&mut request.headers, "Cookie", value)
                }
                _ => {}
            }
            continue;
        }

        if SWITCHES.contains(&flag) {
            continue;
        }
        if token.starts_with('-') && token.len() > 1 {
            // A flag this does not know. What follows is taken for its value
            // unless it looks like another flag or the URL.
            if rest
                .get(index)
                .is_some_and(|next| !next.starts_with('-') && !looks_like_url(next))
            {
                index += 1;
            }
            continue;
        }
        if request.url.is_empty() {
            request.url = token.to_string();
        }
    }

    request.body = data.join("&");
    request.method =
        method.unwrap_or_else(|| if data.is_empty() { "GET" } else { "POST" }.to_string());
    if header(&request.headers, "Content-Type").is_none() {
        if multipart {
            request
                .headers
                .push(("Content-Type".into(), "multipart/form-data".into()));
        } else if form_encoded {
            request.headers.push((
                "Content-Type".into(),
                "application/x-www-form-urlencoded".into(),
            ));
        }
    }
    Ok(request)
}

fn looks_like_url(text: &str) -> bool {
    text.starts_with("http://") || text.starts_with("https://") || text.starts_with("{{")
}

/// What Go's `url.QueryEscape` does, which is what v1 did: unreserved
/// characters as they are, a space as `+`, everything else `%XX`.
fn query_escape(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    for byte in text.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(byte as char)
            }
            b' ' => out.push('+'),
            other => out.push_str(&format!("%{other:02X}")),
        }
    }
    out
}

/// The request as a curl one-liner. `fields` is a multipart body's fields,
/// which go out as `-F` so curl builds the body and its boundary itself.
pub fn format(request: &Request, fields: Option<&[(String, String)]>) -> String {
    let mut out = String::from("curl");
    let method = if request.method.is_empty() {
        "GET"
    } else {
        request.method.as_str()
    };
    if method != "GET" {
        out.push_str(" -X ");
        out.push_str(method);
    }
    for (name, value) in &request.headers {
        if fields.is_some() && name.eq_ignore_ascii_case("Content-Type") {
            continue;
        }
        out.push_str(" -H ");
        out.push_str(&shell_quote(&format!("{name}: {value}")));
    }
    match fields {
        Some(fields) => {
            for (name, value) in fields {
                out.push_str(" -F ");
                out.push_str(&shell_quote(&format!("{name}={value}")));
            }
        }
        None if !request.body.is_empty() => {
            out.push_str(" --data-raw ");
            out.push_str(&shell_quote(&request.body));
        }
        None => {}
    }
    if !request.url.is_empty() {
        out.push(' ');
        out.push_str(&shell_quote(&request.url));
    }
    out
}

/// Single-quotes anything a shell would read as more than text. v1 left `&`
/// bare, so a URL with two query parameters was cut in half by the shell.
fn shell_quote(text: &str) -> String {
    let plain = !text.is_empty()
        && text
            .chars()
            .all(|ch| ch.is_ascii_alphanumeric() || "-_./:=@%+,".contains(ch));
    if plain {
        return text.to_string();
    }
    format!("'{}'", text.replace('\'', r"'\''"))
}

/// Just enough of a POSIX shell to read a pasted command: single and double
/// quotes, `$'…'`, backslash escapes, and `\` at the end of a line.
fn tokenize(text: &str) -> Result<Vec<String>> {
    let unterminated = || Error::Invalid("the curl command has an unterminated quote".into());
    let mut out = Vec::new();
    let mut current = String::new();
    let mut in_token = false;
    let mut chars = text.chars().peekable();

    while let Some(ch) = chars.next() {
        match ch {
            '\'' => {
                in_token = true;
                loop {
                    match chars.next() {
                        Some('\'') => break,
                        Some(inner) => current.push(inner),
                        None => return Err(unterminated()),
                    }
                }
            }
            '"' => {
                in_token = true;
                loop {
                    match chars.next() {
                        Some('"') => break,
                        Some('\\') => match chars.peek() {
                            Some('"' | '\\' | '$' | '`') => {
                                current.push(chars.next().unwrap_or_default())
                            }
                            Some('\n') => {
                                chars.next();
                            }
                            _ => current.push('\\'),
                        },
                        Some(inner) => current.push(inner),
                        None => return Err(unterminated()),
                    }
                }
            }
            '$' if chars.peek() == Some(&'\'') => {
                chars.next();
                in_token = true;
                loop {
                    match chars.next() {
                        Some('\'') => break,
                        Some('\\') => match chars.next() {
                            Some('n') => current.push('\n'),
                            Some('t') => current.push('\t'),
                            Some('r') => current.push('\r'),
                            Some(other) => current.push(other),
                            None => return Err(unterminated()),
                        },
                        Some(inner) => current.push(inner),
                        None => return Err(unterminated()),
                    }
                }
            }
            '\\' => match chars.next() {
                Some('\n') => {}
                Some('\r') => {
                    chars.next_if_eq(&'\n');
                }
                Some(escaped) => {
                    current.push(escaped);
                    in_token = true;
                }
                None => {}
            },
            ch if ch.is_whitespace() => {
                if in_token {
                    out.push(std::mem::take(&mut current));
                    in_token = false;
                }
            }
            ch => {
                current.push(ch);
                in_token = true;
            }
        }
    }
    if in_token {
        out.push(current);
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn recognises_a_curl_command() {
        for (text, expected) in [
            ("curl https://x", true),
            ("  curl https://x", true),
            ("curl", true),
            ("$ curl https://x", true),
            ("wget https://x", false),
            ("https://example.com", false),
        ] {
            assert_eq!(is_curl(text), expected, "{text}");
        }
    }

    #[test]
    fn a_bare_url_is_a_get() {
        let request = parse("curl https://example.com/x").unwrap();
        assert_eq!(
            (request.method.as_str(), request.url.as_str()),
            ("GET", "https://example.com/x")
        );
    }

    #[test]
    fn reads_method_headers_and_data() {
        let request = parse(
            r#"curl -X POST -H "Content-Type: application/json" -H 'Authorization: Bearer abc' --data-raw '{"a":1}' https://example.com/api"#,
        )
        .unwrap();
        assert_eq!(request.method, "POST");
        assert_eq!(request.url, "https://example.com/api");
        assert_eq!(request.header("Content-Type"), Some("application/json"));
        assert_eq!(request.header("Authorization"), Some("Bearer abc"));
        assert_eq!(request.body, r#"{"a":1}"#);
    }

    #[test]
    fn data_implies_a_form_post() {
        let request = parse(r#"curl -d "foo=bar" https://x"#).unwrap();
        assert_eq!(request.method, "POST");
        assert_eq!(
            request.header("Content-Type"),
            Some("application/x-www-form-urlencoded")
        );
    }

    #[test]
    fn keeps_every_field_of_a_client_credentials_request() {
        let request = parse(
            "curl -s -X POST 'https://login.microsoftonline.com/tid/oauth2/v2.0/token' \
             -d 'grant_type=client_credentials' \
             -d 'client_id=abc' \
             -d 'scope=https://graph.microsoft.com/.default' \
             --data-urlencode 'client_secret=sec.ret~with/chars'",
        )
        .unwrap();
        assert_eq!(request.method, "POST");
        for expected in [
            "grant_type=client_credentials",
            "client_id=abc",
            "scope=https://graph.microsoft.com/.default",
            "client_secret=sec.ret~with%2Fchars",
        ] {
            assert!(
                request.body.contains(expected),
                "{expected} missing from {}",
                request.body
            );
        }
    }

    #[test]
    fn user_becomes_basic_auth() {
        let request = parse("curl -u alice:secret https://x").unwrap();
        assert_eq!(
            request.header("Authorization"),
            Some("Basic YWxpY2U6c2VjcmV0")
        );
    }

    #[test]
    fn follows_line_continuations() {
        let request = parse("curl -X POST \\\n  -H 'A: 1' \\\n  https://x").unwrap();
        assert_eq!(request.url, "https://x");
        assert_eq!(request.header("A"), Some("1"));
    }

    #[test]
    fn reads_what_a_browser_copies() {
        let request = parse(
            "curl 'https://api.example.com/graphql' \
             -H 'accept: */*' \
             -b 'session=abc; theme=dark' \
             --data-raw $'{\"q\":\"it\\'s\"}' \
             --compressed",
        )
        .unwrap();
        assert_eq!(request.url, "https://api.example.com/graphql");
        assert_eq!(request.header("Cookie"), Some("session=abc; theme=dark"));
        assert_eq!(request.body, r#"{"q":"it's"}"#);
        assert_eq!(request.method, "POST");
    }

    #[test]
    fn reads_values_written_against_their_flag() {
        let request = parse("curl -XPUT -H'X-A: 1' https://x").unwrap();
        assert_eq!(request.method, "PUT");
        assert_eq!(request.header("X-A"), Some("1"));
    }

    #[test]
    fn an_unterminated_quote_is_an_error() {
        assert!(parse("curl 'https://x").is_err());
    }

    #[test]
    fn exports_what_the_shell_will_read_back() {
        let request = Request {
            method: "POST".into(),
            url: "https://x/search?a=1&b=2".into(),
            headers: vec![("Content-Type".into(), "application/json".into())],
            body: r#"{"q":"it's"}"#.into(),
            ..Request::default()
        };
        let command = format(&request, None);
        assert_eq!(
            command,
            r#"curl -X POST -H 'Content-Type: application/json' --data-raw '{"q":"it'\''s"}' 'https://x/search?a=1&b=2'"#
        );
        let again = parse(&command).unwrap();
        assert_eq!(again.url, request.url);
        assert_eq!(again.body, request.body);
    }

    #[test]
    fn a_multipart_body_exports_as_fields() {
        let request = Request {
            method: "POST".into(),
            url: "https://x/upload".into(),
            headers: vec![(
                "Content-Type".into(),
                "multipart/form-data; boundary=b".into(),
            )],
            ..Request::default()
        };
        let fields = vec![("file".to_string(), "@/tmp/a.png".to_string())];
        assert_eq!(
            format(&request, Some(&fields)),
            "curl -X POST -F file=@/tmp/a.png https://x/upload"
        );
    }
}
